File: tools/dbench/answers/go_haiku/s2/m2_cfg_chain.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func main() {
	data, _ := os.ReadFile("app.cfg")
	var m map[string]interface{}
	json.Unmarshal(data, &m)

	mode := "prod"
	if d, ok := m["debug"].(bool); ok && d {
		mode = "debug"
	}

	host := "localhost"
	if h, ok := m["host"].(string); ok && h != "" {
		host = h
	}

	port := 0
	if p, ok := m["port"].(float64); ok {
		port = int(p)
	}

	fmt.Printf("%s %s:%d\n", mode, host, port)
}

File: tools/dbench/answers/rust_haiku/s2/m2_cfg_chain.rs
use std::fs;

fn get_bool(json: &str, key: &str) -> Option<bool> {
    let pat = format!("\"{}\":", key);
    if let Some(pos) = json.find(&pat) {
        let rest = &json[pos + pat.len()..].trim_start();
        if rest.starts_with("true") {
            Some(true)
        } else if rest.starts_with("false") {
            Some(false)
        } else {
            None
        }
    } else {
        None
    }
}

fn get_str(json: &str, key: &str) -> Option<String> {
    let pat = format!("\"{}\":", key);
    if let Some(pos) = json.find(&pat) {
        let rest = &json[pos + pat.len()..].trim_start();
        if rest.starts_with('"') {
            let val = &rest[1..];
            if let Some(end) = val.find('"') {
                return Some(val[..end].to_string());
            }
        }
    }
    None
}

fn main() {
    let cfg = fs::read_to_string("app.cfg").unwrap_or_default();

    let is_debug = get_bool(&cfg, "debug").unwrap_or(false);
    let mode = if is_debug { "debug" } else { "prod" };

    let host = get_str(&cfg, "host").unwrap_or_else(|| "localhost".to_string());
    let port = get_str(&cfg, "port").unwrap_or_else(|| "8080".to_string());

    println!("{} {}:{}", mode, host, port);
}
